docs/middleware.md

```diff
@@ -0,0 +1,218 @@
+# Middleware
+
+Kstreams allows you to include middlewares for adding behavior to streams. 
+
+A *middleware* is a `callable` that works with every `ConsumerRecord` (CR) *before* it is processed by a specific `stream`.  Also *after* the `CR` has been handled `Middlewares` also have access to the `stream`, `send` function.
+
+- It takes each `CR` that arrives to a topic.
+- It can then do something to the `CR` or run any needed code.
+- Then it passes the `CR` to be processed by the specific stream.
+- Once the `CR` is processed by the stream, the chain is "completed".
+- If there is code after the `self.next_call(cr)` then it will be executed.
+
```

Comment:
```suggestion
Let's say we assign the following Middlewares: `[MiddlewareFoo, MiddlewareBar]`

We can see that `MiddlewareFoo` will be called first, and last:

```mermaid
flowchart LR
    m1i[MiddlewareFoo]
    m2i[MiddlewareBar]
    fn[stream]
    m1o[MiddlewareFoo]
    m2o[MiddlewareBar]
    
    m1i --> m2i
    m2i --> fn
    fn --> m2o
    m2o --> m1o
    ```
```

Reply:
I would add something like that, maybe better phrased

Reply:
What you point is the description of `middleware chain`. I think at the beginning of the docs is better to describe the process as simple as possible focus in the ConsumerRecord, `Stream` and  _one_ `middleware`. I will addyour suggestion in the `Middleware chain` section

Reply:
After discussing we realized that a sequence diagram is a better option
![Screenshot 2024-01-16 at 16 16 51](https://github.com/kpn/kstreams/assets/2828842/c0e7dfe2-34b2-49a1-9d58-5a1233efec54)